urld025f21ac3_mvattipulusu_Innovation2013_tgz-pJ8-com_usps_data_ManifestFileTestJ8/src/com/usps/evs/vo/PicCode.java

```diff
@@ -0,0 +1,319 @@
+				
+				System.out.println("Into international scan I01 parsing:"+pic.toString());
+				String appID = pic.substring(0, 2);
+				System.out.println("Into international scan parsing:"+appID);
+				System.out.println("Into international scan parsing- package id:"+pic.substring(2,10));
+				System.out.println("Into international scan parsing - checkdigit:"+pic.substring(10,11));
+		
+				//I01 barcodes
+				setAppicationID(pic.substring(0,2));
+				setPackageID(pic.substring(2, 10));
+				setCheckDigit(pic.substring(10, 11));
+				setDunsPkgidDzip(pic);
+				
+			}
+			else if (pic.length() <= 26 && pic.length() >= 18){
+				setAppicationID(pic.substring(0,2));
+				setDunsPkgidDzip(pic);
+				
+			}			
+			else	{
+				setServiceTypeCode(null);
+				setDuns(null);
+				setPackageID(null);
+				setCheckDigit(null);
+				setZip(null);
+				setDunsPkgidDzip(null);
+			}
+			
+			if(getDunsPkgidDzip() != null){
+				//Process the DunsPKID to get populate the PIC
+				if("91".equals(getAppicationID()))
+					parseAppID91();
+				else if("92".equals(getAppicationID()))
+					parseAppID92();
+				else if("93".equals(getAppicationID()))
+					parseAppID93();		
+			}
+
+
+		} else {
+			if(pic.length() <= 30 &&
+				pic.length() >= 24)
+			{
+				setServiceTypeCode(pic.substring(10, 12));
+				setDuns(pic.substring(12, 21));
+				setPackageID(pic.substring(21, pic.length() - 1));
+				setCheckDigit(pic.substring(pic.length() - 1, pic.length()));
+				setZip(pic.substring(3, 8));
+				setDunsPkgidDzip(pic.substring(8));
+			}
+			else if(pic.length() <= 22 &&
+				pic.length() >= 18)
+			{
+				setServiceTypeCode(pic.substring(2, 4));
+				setDuns(pic.substring(4, 13));
+				setPackageID(pic.substring(13, pic.length() - 1));
+				setCheckDigit(pic.substring(pic.length() - 1, pic.length()));
+				setZip(null);
+				setDunsPkgidDzip(pic);
+			}
+			else if(pic.length() <= 20 && pic.length() >= 16)
+			{
+				setServiceTypeCode(pic.substring(0, 2));
+				setDuns(pic.substring(2, 11));
+				if(pic.length() == 20)
+				{
+					setPackageID(pic.substring(11, 19));
+					setCheckDigit(pic.substring(19, 20));
+				}
+				else
+				{
+					setPackageID(pic.substring(11, pic.length() - 1));
+					setCheckDigit(pic.substring(pic.length() - 1, pic.length()));
+				}
+				setZip(null);
+			}
+			setDunsPkgidDzip(pic);
+		}	
+    }
+	
+	public boolean parseAppID91(){
+		String pic = getDunsPkgidDzip(); 
+		setServiceTypeCode(pic.substring(2, 4));
+		setDuns(pic.substring(4, 13));
+		setPackageID(pic.substring(13, pic.length() - 1));
+		setCheckDigit(pic.substring(pic.length() - 1, pic.length()));
+		setZip(null);
+
+		return true;
+	}
+
+	public boolean parseAppID92(){
+		String pic = getDunsPkgidDzip(); 
+		 
+		setServiceTypeCode(pic.substring(2, 5));
+		setDuns(pic.substring(5, 14));
+		setPackageID(pic.substring(14, pic.length() - 1));
+		setCheckDigit(pic.substring(pic.length() - 1, pic.length()));
+
+		return true;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 219:226, 230:237.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.